orc/src/main/java/org/apache/iceberg/orc/ExpressionToSearchArgument.java

```diff
@@ -0,0 +1,355 @@
+    return () -> this.builder.startNot()
+          .in(idToColumnName.get(expr.ref().fieldId()), type(expr.ref().type()),
+              literalSet.stream().map(lit -> literal(expr.ref().type(), lit)).toArray(Object[]::new))
+          .end();
+  }
+
+  @Override
+  public <T> Action startsWith(Bound<T> expr, Literal<T> lit) {
+    // Cannot push down STARTS_WITH operator to ORC, so return TruthValue.YES_NO_NULL which signifies
+    // that this predicate cannot help with filtering
+    return () -> this.builder.literal(TruthValue.YES_NO_NULL);
+  }
+
+  @Override
+  public <T> Action predicate(BoundPredicate<T> pred) {
+    if (UNSUPPORTED_TYPES.contains(pred.ref().type().typeId())) {
+      // Cannot push down predicates for types which cannot be represented in PredicateLeaf.Type, so return
+      // TruthValue.YES_NO_NULL which signifies that this predicate cannot help with filtering
+      return () -> this.builder.literal(TruthValue.YES_NO_NULL);
+    } else {
+      return super.predicate(pred);
+    }
+  }
+
+  @FunctionalInterface
+  interface Action {
+    void invoke();
+  }
+
+  private PredicateLeaf.Type type(Type icebergType) {
+    switch (icebergType.typeId()) {
+      case BOOLEAN:
+        return PredicateLeaf.Type.BOOLEAN;
+      case INTEGER:
+      case LONG:
+      case TIME:
+        return PredicateLeaf.Type.LONG;
+      case FLOAT:
+      case DOUBLE:
+        return PredicateLeaf.Type.FLOAT;
+      case DATE:
+        return PredicateLeaf.Type.DATE;
+      case TIMESTAMP:
+        return PredicateLeaf.Type.TIMESTAMP;
+      case STRING:
+        return PredicateLeaf.Type.STRING;
+      case DECIMAL:
+        return PredicateLeaf.Type.DECIMAL;
+      default:
+        throw new UnsupportedOperationException("Type " + icebergType + " not supported in ORC SearchArguments");
+    }
+  }
+
+  private <T> Object literal(Type icebergType, T icebergLiteral) {
+    switch (icebergType.typeId()) {
+      case BOOLEAN:
+      case LONG:
+      case TIME:
+      case DOUBLE:
+        return icebergLiteral;
+      case INTEGER:
+        return ((Integer) icebergLiteral).longValue();
+      case FLOAT:
+        return ((Float) icebergLiteral).doubleValue();
+      case STRING:
+        return icebergLiteral.toString();
+      case DATE:
+        return Date.valueOf(LocalDate.ofEpochDay((Integer) icebergLiteral));
+      case TIMESTAMP:
+        long microsFromEpoch = (Long) icebergLiteral;
+        return Timestamp.from(Instant.ofEpochSecond(Math.floorDiv(microsFromEpoch, 1_000_000),
+            (microsFromEpoch % 1_000_000) * 1_000));
+      case DECIMAL:
+        return new HiveDecimalWritable(HiveDecimal.create((BigDecimal) icebergLiteral, false));
+      default:
+        throw new UnsupportedOperationException("Type " + icebergType + " not supported in ORC SearchArguments");
+    }
+  }
+
+  /**
+   * Generates mapping from field IDs to fully qualified column names compatible with ORC convention for a given
+   * {@link Schema}
```

Comment:
Nit: missing `.` at the end of this sentence. This is also a bit long for a description. What about "Generates a map from field ID to ORC qualified name."